src/main/java/de/eldoria/pickmeup/listener/CarryListener.java

```diff
@@ -92,57 +92,113 @@ public void onEntityInteract(PlayerInteractAtEntityEvent event) {
+        switch (currentState) {
+            case SNEAK_MOUNT:
+                mountStates.put(playerUUID, MountState.WALKING);
+                break;
+            case SNEAK_THROW:
+                if (!throwBarHandler.isRegistered(player)) {
+                    unmountAll(player);
+                } else {
+                    performThrow(player, playerUUID);
+                }
+                break;
         }
+    }
 
-        if (event.isSneaking() && mountState == MountState.WALKING) {
-            mountStates.put(player.getUniqueId(), MountState.SNEAK_THROW);
+    private void handleSneakPress(Player player, UUID playerUUID, MountState currentState) {
+        if (currentState == MountState.WALKING) {
+            mountStates.put(playerUUID, MountState.SNEAK_THROW);
             delayedActions.schedule(() -> {
-                        if (player.isSneaking()) {
-                            throwBarHandler.register(player);
-                        }
-                    }
-                    , 10);
-            return;
+                if (player.isSneaking() && mountStates.get(playerUUID) == MountState.SNEAK_THROW) {
+                    throwBarHandler.register(player);
+                }
+            }, 10);
         }
+    }
 
-        if (!event.isSneaking() && mountState == MountState.SNEAK_THROW) {
-            if (!throwBarHandler.isRegistered(player)) {
-                unmountAll(player);
-            } else {
-                double force = throwBarHandler.getAndRemove(player);
-                Vector throwVec = player.getEyeLocation().getDirection().normalize().multiply(force * config.carrySettings().throwForce());
-                player.getWorld().playSound(player.getLocation(), Sound.ENTITY_ENDER_DRAGON_FLAP, 1, 1);
-                for (Entity passenger : player.getPassengers()) {
-                    delayedActions.schedule(() -> trailHandler.startTrail(passenger), 2);
-                    player.removePassenger(passenger);
-                    passenger.setVelocity(throwVec);
-                    plugin.getLogger().config("Throwing entity | Location:" + player.getLocation().toVector()
-                                              + " | Force: " + force
-                                              + " | ThrowForce: " + config.carrySettings().throwForce()
-                                              + " | ViewVec: " + throwVec);
+    private void performThrow(Player player, UUID playerUUID) {
+        double force = throwBarHandler.getAndRemove(player);
+        Vector throwVec = player.getEyeLocation().getDirection().normalize().multiply(force * config.carrySettings().throwForce());
+        player.getWorld().playSound(player.getLocation(), Sound.ENTITY_ENDER_DRAGON_FLAP, 1, 1);
+
+        List<Entity> passengers = new ArrayList<>(player.getPassengers());
+        for (Entity passenger : passengers) {
+            if (passenger == null || !passenger.isValid() || passenger.isDead()) {
```

Comment:
An entity is not valid it is dead:
https://jd.papermc.io/paper/1.21.7/org/bukkit/entity/Entity.html#isValid()
Returns false if the entity has died, been despawned for some other reason, or has not been added to the world.

```suggestion
            if (passenger == null || !passenger.isValid()) {
```